spec/callbacks_spec.rb

```diff
@@ -48,21 +56,34 @@ def method_missing(method, *args)
 end
 
 describe CallbackCloud do
-  before { @fs = CallbackCloud.new(File.dirname(__FILE__) + '/files', 'http://assets/') }
+  before do
+    @fs = CallbackCloud.new(File.dirname(__FILE__) + '/files', 'http://assets/')
+    @fs.write('tmp/file.txt', 'foo')
+  end
 
   it "should invoke callbacks after store" do
     @fs.should_receive(:callback_before_write).with('tmp/file.txt', 'text').and_return(true)
     @fs.should_receive(:callback_after_write).with('tmp/file.txt', 'text').and_return(true)
 
 
-    @fs.write 'tmp/file.txt', 'text'
+    @fs.write('tmp/file.txt', 'text').should == true
+    @fs.read('tmp/file.txt').should == 'text'
   end
 
   it "should invoke callbacks after delete" do
     @fs.should_receive(:callback_before_delete).with('tmp/file.txt').and_return(true)
     @fs.should_receive(:callback_after_delete).with('tmp/file.txt').and_return(true)
 
-    @fs.delete 'tmp/file.txt'
+    @fs.delete('tmp/file.txt').should == 'foo'
+  end
+
+  it "should not invoke other callbacks when a before_ filter returns false" do
+    @fs.should_receive(:callback_before_delete)
+      .with('tmp/file.txt')
+      .and_return(false)
+    @fs.should_not_receive(:callback_after_delete)
+
+    @fs.delete('tmp/file.txt').should == nil
```

Comment:
Should we make this return `false`, to be more explicit?

Reply:
I could look at that separately; namely I'd like to contain the changes necessary and audit what the consumers expect from the APIs... it will entail to change the return value to an explicit boolean from all the bucket implementations. Some [return the value tied to the key or `nil`](https://github.com/Shopify/asset_cloud/blob/master/lib/asset_cloud/buckets/memory_bucket.rb#L21-L23). Whereas others [have the return value set regardless of the success or failure of the operation](https://github.com/Shopify/asset_cloud/blob/master/lib/asset_cloud/buckets/s3_bucket.rb#L26-L32).